internal/workerpool/pool.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2018 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package workerpool provides a func called Run that allows doing work
+// concurrently, only spawning goroutines as needed and not spawning more of
+// them than a set limit. The idea for this package comes from Bryan C. Mills's
+// talk: https://www.youtube.com/watch?v=5zXAHh5tJqQ&t=33m13s
+package workerpool
+
+import (
+	"context"
+)
+
+// Run runs a worker pool with no more than limit goroutines. It gets tasks
+// from the nextTask func, which returns a task and a boolean that tells
+// whether to work on the task and continue running the pool. The doWork func
+// processes a single task. The provided context can be used to cancel
+// everything and exit the loop.
+func Run(ctx context.Context, limit int, nextTask func(context.Context) (interface{}, bool), doWork func(ctx context.Context, task interface{})) {
+	type token struct{}
+	sem := make(chan token, limit)
```

Comment:
It looks like this takes constant space, no matter what limit is. That's good.

Reply:
Yep.